internal/controllers/provisioningrequest_controller.go

```diff
@@ -224,42 +224,50 @@ func (t *provisioningRequestReconcilerTask) run(ctx context.Context) (ctrl.Resul
 	if err != nil {
 		return requeueWithError(err)
 	}
+	if !utils.IsClusterProvisionPresent(t.object) ||
+		utils.IsClusterProvisionTimedOutOrFailed(t.object) {
+		return doNotRequeue(), nil
+	}
 
 	// Handle policy configuration only after the cluster provisioning
-	// has started, and not failed or timedout (completed, in-progress or unknown)
-	if utils.IsClusterProvisionPresent(t.object) &&
-		!utils.IsClusterProvisionTimedOutOrFailed(t.object) {
-
-		// Handle configuration through policies.
-		requeue, err := t.handleClusterPolicyConfiguration(ctx)
-		if err != nil {
-			return requeueWithError(err)
-		}
-
-		// Requeue if cluster provisioning is not completed (in-progress or unknown)
-		// or there are enforce policies that are not Compliant.
-		if !utils.IsClusterProvisionCompleted(t.object) || requeue {
-			return requeueWithLongInterval(), nil
-		}
+	// has started, and not failed or timedout
+	requeueForConfig, err := t.handleClusterPolicyConfiguration(ctx)
+	if err != nil {
+		return requeueWithError(err)
+	}
 
+	if utils.IsClusterZtpDone(t.object) {
+		// If the initial provisioning is completed, check if an upgrade is requested
 		shouldUpgrade, err := t.IsUpgradeRequested(ctx, renderedClusterInstance.GetName())
 		if err != nil {
 			return requeueWithError(err)
 		}
 
-		if utils.IsClusterUpgradeInitiated(t.object) && !utils.IsClusterUpgradeCompleted(t.object) ||
-			utils.IsClusterProvisionCompleted(t.object) && shouldUpgrade {
-			t.logger.InfoContext(
-				ctx,
-				"Upgrade requested. Start handling upgrade.",
-			)
-			requeue, err := t.handleUpgrade(ctx, renderedClusterInstance)
+		// An upgrade is requested or upgrade has started but not completed
+		if shouldUpgrade ||
+			(utils.IsClusterUpgradeInitiated(t.object) &&
+				!utils.IsClusterUpgradeCompleted(t.object)) {
+			upgradeCtrlResult, err := t.handleUpgrade(ctx, renderedClusterInstance.GetName())
 			if err != nil {
 				return requeueWithError(err)
 			}
-			return requeue, nil
+			if !upgradeCtrlResult.IsZero() {
```

Comment:
Would it be the same here to use `if upgradeCtrlResult != doNotRequeue()`? I think it would be more clear, but it's up to you.
A comment above explaining the use case a bit more would definitely be helpful. :pray:

Reply:
yup, sure. I will change to `upgradeCtrlResult.RequeueAfter > 0` which seems to be more clear